assets/js/data/cart/push-changes.ts

```diff
@@ -1,284 +1,181 @@
+	doingPush: false,
+	// Local cache of the last pushed customerData used for comparisons.
+	customerData: {
+		billingAddress: {} as CartBillingAddress,
+		shippingAddress: {} as CartShippingAddress,
+	},
+	// Tracks which props have changed so the correct data gets pushed to the server.
+	dirtyProps: {
+		billingAddress: [] as BaseAddressKey[],
+		shippingAddress: [] as BaseAddressKey[],
+	},
 };
 
 /**
- * Checks if a cart response contains an email property.
+ * Initializes the customer data cache on the first run.
  */
-const isBillingAddress = (
-	address: BillingOrShippingAddress
-): address is CartBillingAddress => {
-	return 'email' in address;
+const initialize = () => {
+	localState.customerData = select( STORE_KEY ).getCustomerData();
+	localState.customerDataIsInitialized = true;
 };
 
 /**
- * Trims and normalizes address data for comparison.
+ * Checks customer data against new customer data to get a list of dirty props.
  */
-export const normalizeAddress = ( address: BillingOrShippingAddress ) => {
-	const trimmedAddress = Object.entries( address ).reduce(
-		( acc, [ key, value ] ) => {
-			//Skip normalizing for any non string field
-			if ( typeof value !== 'string' ) {
-				acc[ key as keyof BillingOrShippingAddress ] = value;
-				return acc;
-			}
-
-			if ( key === 'postcode' ) {
-				acc[ key as keyof BillingOrShippingAddress ] = value
-					.replace( ' ', '' )
-					.toUpperCase();
-				return acc;
-			}
+const updateDirtyProps = () => {
+	// Returns all current customer data from the store.
+	const newCustomerData = select( STORE_KEY ).getCustomerData();
+
+	localState.dirtyProps.billingAddress = [
+		...localState.dirtyProps.billingAddress,
+		...getDirtyKeys(
+			localState.customerData.billingAddress,
+			newCustomerData.billingAddress
+		),
+	];
+
+	localState.dirtyProps.shippingAddress = [
+		...localState.dirtyProps.shippingAddress,
+		...getDirtyKeys(
+			localState.customerData.shippingAddress,
+			newCustomerData.shippingAddress
+		),
+	];
 
-			acc[ key as keyof BillingOrShippingAddress ] = value.trim();
-			return acc;
-		},
-		{} as BillingOrShippingAddress
-	);
-	return trimmedAddress;
+	// Update local cache of customer data so the next time this runs, it can compare against the latest data.
+	localState.customerData = newCustomerData;
 };
 
 /**
- * Does a shallow compare of all address data to determine if the cart needs updating on the server.
+ * Function to dispatch an update to the server.
  */
-const isAddressDirty = < T extends CartBillingAddress | CartShippingAddress >(
-	// An object containing all previous address information
-	previousAddress: T,
-	// An object containing all address information.
-	address: T
-): boolean => {
-	if (
-		isBillingAddress( address ) &&
-		pluckEmail( address ) !==
-			pluckEmail( previousAddress as CartBillingAddress )
-	) {
-		return true;
+const updateCustomerData = (): void => {
+	if ( localState.doingPush ) {
+		return;
```

Comment:
imo instead of blocking pushes, in the future, its best to cancel the current push and let the new one go. cleaning up dirty props happens after we're done pushing.